File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/PushNotifications/MyFirebaseMessagingService.java
package com.digicoffer.lauditor.CommonFiles.PushNotifications;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.media.RingtoneManager;
import android.net.Uri;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import com.digicoffer.lauditor.MainActivity;
import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import org.json.JSONObject;

public class MyFirebaseMessagingService extends FirebaseMessagingService implements AsyncTaskCompleteListener {

    private static final String TAG = "FCM_Service";
    private static final String CHANNEL_ID = "lauditor_channel";
    private static final String CHANNEL_NAME = "Lauditor Notifications";
    private static final String PREFS_NAME = "MyPrefs";
    private static final String TOKEN_KEY = "fcm_token";

    @Override
    public void onNewToken(@NonNull String token) {
        super.onNewToken(token);
        Log.d(TAG, "New FCM Token Generated: " + token);
        saveTokenLocally(token);
        sendTokenToServer(token);
    }

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);

        String title = "";
        String body = "";
        String navigation = "";

        if (remoteMessage.getNotification() != null) {
            title = remoteMessage.getNotification().getTitle() != null
                    ? remoteMessage.getNotification().getTitle() : "";
            body = remoteMessage.getNotification().getBody() != null
                    ? remoteMessage.getNotification().getBody() : "";
        }

        if (!remoteMessage.getData().isEmpty()) {
            if (remoteMessage.getData().get("title") != null) {
                title = remoteMessage.getData().get("title");
            }
            if (remoteMessage.getData().get("body") != null) {
                body = remoteMessage.getData().get("body");
            }
            if (remoteMessage.getData().get("navigation") != null) {
                navigation = remoteMessage.getData().get("navigation");
            }
        }
        // Push notification for chat disabled — skip if this is a chat notification
        try {
            if (navigation != null && !navigation.isEmpty()) {
                JSONObject navJson = new JSONObject(navigation);
                String routeName = navJson.optString("route_name", "");
                if ("message_client_inbox".equalsIgnoreCase(routeName)) {
                    Log.d(TAG, "onMessageReceived — chat notification suppressed");
                    return;
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "onMessageReceived — failed to parse navigation JSON: " + e.getMessage());
        }

        if (!title.isEmpty() || !body.isEmpty()) {
            showNotification(title, body, navigation);
        }
    }
//    @Override
//    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
//        super.onMessageReceived(remoteMessage);
//
//        String title = "";
//        String body = "";
//        String navigation = ""; // ← the full navigation JSON string
//
//        if (remoteMessage.getNotification() != null) {
//            title = remoteMessage.getNotification().getTitle() != null
//                    ? remoteMessage.getNotification().getTitle() : "";
//            body = remoteMessage.getNotification().getBody() != null
//                    ? remoteMessage.getNotification().getBody() : "";
//        }
//
//        if (!remoteMessage.getData().isEmpty()) {
//            if (remoteMessage.getData().get("title") != null) {
//                title = remoteMessage.getData().get("title");
//            }
//            if (remoteMessage.getData().get("body") != null) {
//                body = remoteMessage.getData().get("body");
//            }
//            // ── Pull the full navigation JSON string from data payload ─
//            if (remoteMessage.getData().get("navigation") != null) {
//                navigation = remoteMessage.getData().get("navigation");
//            }
//        }
//
//        if (!title.isEmpty() || !body.isEmpty()) {
//            showNotification(title, body, navigation); // ← 3 args now
//        }
//    }

    // ── Updated to 3 params — navigation replaces screen + itemId ────────
    private void showNotification(String title, String body, String navigation) {

        Intent intent = new Intent(this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
                | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);

        // Pass the full navigation JSON — MainActivity will parse it
        if (navigation != null && !navigation.isEmpty()) {
            intent.putExtra("fcm_navigation", navigation);
        }

        int requestCode = (int) System.currentTimeMillis();
        PendingIntent pendingIntent = PendingIntent.getActivity(
                this,
                requestCode,
                intent,
                PendingIntent.FLAG_ONE_SHOT | PendingIntent.FLAG_IMMUTABLE
        );

        Uri soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);

        NotificationManager manager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        if (manager == null) {
            Log.e(TAG, "NotificationManager is null — cannot show notification");
            return;
        }

        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_HIGH
        );
        channel.setDescription("Push notifications for LexiZLawyers App");
        channel.enableLights(true);
        channel.setLightColor(Color.GREEN);
        channel.enableVibration(true);
        channel.setVibrationPattern(new long[]{0, 500, 200, 500});
        channel.setShowBadge(true);
        channel.setLockscreenVisibility(NotificationCompat.VISIBILITY_PUBLIC);
        manager.createNotificationChannel(channel);

        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(this, CHANNEL_ID)
                        .setSmallIcon(R.mipmap.ic_launcher)
                        .setContentTitle(title)
                        .setContentText(body)
                        .setStyle(new NotificationCompat.BigTextStyle()
                                .bigText(body)
                                .setBigContentTitle(title))
                        .setAutoCancel(true)
                        .setPriority(NotificationCompat.PRIORITY_HIGH)
                        .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                        .setSound(soundUri)
                        .setDefaults(NotificationCompat.DEFAULT_VIBRATE)
                        .setLights(Color.GREEN, 1000, 500)
                        .setBadgeIconType(NotificationCompat.BADGE_ICON_SMALL)
                        .setContentIntent(pendingIntent);

        int notificationId = (int) System.currentTimeMillis();
        manager.notify(notificationId, builder.build());

        Log.d(TAG, "Notification displayed with ID: " + notificationId);
    }

    private void saveTokenLocally(String token) {
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit()
                .putString(TOKEN_KEY, token)
                .apply();
        Log.d(TAG, "FCM Token saved to SharedPreferences ✅");
    }

    private void sendTokenToServer(String token) {
        try {
            JSONObject postData = new JSONObject();
            postData.put("userId", Constants.USER_ID);
            postData.put("fcmToken", token);
            postData.put("platform", "android");

            // ✅ Correct parameter order for your WebServiceHelper:
            // callHttpWebService(callback, context, methodType, url, requestType, body)
            WebServiceHelper.callHttpWebService(
                    null, this,                                      // context
                    WebServiceHelper.RestMethodType.POST,      // REST method
                    Constants.Notification_Base_Url + "/register-token",                       // ← Replace with your actual API endpoint
                    "FCM_TOKEN_UPDATE",                        // request type tag — handled in onAsyncTaskComplete()
                    postData.toString()                        // POST body as JSON string
            );

            Log.d("FCM", "FCM Token sent to backend ✅");

        } catch (Exception e) {
            Log.e("FCM", "Failed to send FCM token to backend: "
                    + e.getMessage());
            e.printStackTrace();
        }
    }

    // Add this method to MyFirebaseMessagingService class
    public static void clearAllNotifications(Context context) {
        NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (manager != null) {
            manager.cancelAll();
            Log.d(TAG, "All notifications cleared from notification center ✅");
        }
    }

    public static void logoutToken(Context context, String token) {
        try {
            JSONObject postData = new JSONObject();
            postData.put("fcmToken", token);

            WebServiceHelper.callHttpWebService(
                    null, context,
                    WebServiceHelper.RestMethodType.POST,
                    Constants.Notification_Base_Url + "/logout-token",
                    "FCM_TOKEN_LOGOUT",
                    postData.toString()
            );

            Log.d("FCM", "Logout token sent to backend ✅");
        } catch (Exception e) {
            Log.e("FCM", "Failed to send logout token to backend: " + e.getMessage());
        }
    }

    @Override
    public void onDeletedMessages() {
        super.onDeletedMessages();
        Log.d(TAG, "Messages deleted from FCM server");
    }

    @Override
    public void onSendError(@NonNull String msgId, @NonNull Exception exception) {
        super.onSendError(msgId, exception);
        Log.e(TAG, "Message send error — ID: " + msgId + " Error: " + exception.getMessage());
    }

    @Override
    public void onMessageSent(@NonNull String msgId) {
        super.onMessageSent(msgId);
        Log.d(TAG, "Message sent successfully — ID: " + msgId);
    }

    @Override
    public void onClick(View view) {

    }

    @Override
    public void onAsyncTaskComplete(HttpResultDo httpResult) {
        try {
            if (httpResult.getResult() == WebServiceHelper.ServiceCallStatus.Success) {

                JSONObject result = new JSONObject(httpResult.getResponseContent());
                if (httpResult.getRequestType().equals("FCM_TOKEN_UPDATE")) {
                    boolean error = result.optBoolean("error");
                    String msg = result.optString("message");
                    Log.d("FCM_TOKEN_UPDATE", msg);
                }
            }
        } catch (Exception e) {
            e.fillInStackTrace();
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/CommonFiles/PushNotifications/MyFirebaseMessagingService.kt
package com.digicoffer.lauditor.CommonFiles.PushNotifications

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.media.AudioAttributes
import android.media.RingtoneManager
import android.util.Log
import android.view.View
import androidx.core.app.NotificationCompat
import com.digicoffer.lauditor.MainActivity
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import org.json.JSONObject

class MyFirebaseMessagingService : FirebaseMessagingService(), AsyncTaskCompleteListener {

    override fun onNewToken(token: String) {
        super.onNewToken(token)
        Log.d(TAG, "New FCM Token Generated: $token")
        saveTokenLocally(token)
        sendTokenToServer(token)
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        val startTime = System.currentTimeMillis()
        Log.d("PUSH_DEBUG", "PUSH_DEBUG | onMessageReceived START | thread = ${Thread.currentThread().name} | timestamp = $startTime")

        val parseStart = System.currentTimeMillis()
        var title = ""
        var body = ""
        var navigation = ""

        remoteMessage.notification?.let {
            title = it.title ?: ""
            body = it.body ?: ""
        }

        if (remoteMessage.data.isNotEmpty()) {
            remoteMessage.data["title"]?.let { title = it }
            remoteMessage.data["body"]?.let { body = it }
            remoteMessage.data["navigation"]?.let { navigation = it }
        }
        val parseDuration = System.currentTimeMillis() - parseStart
        Log.d("PUSH_DEBUG", "PUSH_DEBUG | parsePayload END | duration = ${parseDuration}ms | title = $title | navigation = $navigation")

        if (title.isNotEmpty() || body.isNotEmpty()) {
            val buildStart = System.currentTimeMillis()
            showNotification(title, body, navigation)
            val buildDuration = System.currentTimeMillis() - buildStart
            Log.d("PUSH_DEBUG", "PUSH_DEBUG | showNotification END | duration = ${buildDuration}ms")
        }

        val totalDuration = System.currentTimeMillis() - startTime
        Log.d("PUSH_DEBUG", "PUSH_DEBUG | onMessageReceived END | totalDuration = ${totalDuration}ms")
    }

    private fun showNotification(title: String, body: String, navigation: String?) {
        val intent = Intent(this, MainActivity::class.java).apply {
            addFlags(
                Intent.FLAG_ACTIVITY_CLEAR_TOP or
                        Intent.FLAG_ACTIVITY_SINGLE_TOP or
                        Intent.FLAG_ACTIVITY_NEW_TASK
            )
        }

        if (!navigation.isNullOrEmpty()) {
            intent.putExtra("fcm_navigation", navigation)
        }

        val requestCode = System.currentTimeMillis().toInt()
        val pendingIntent = PendingIntent.getActivity(
            this,
            requestCode,
            intent,
            PendingIntent.FLAG_ONE_SHOT or PendingIntent.FLAG_IMMUTABLE
        )

        initNotificationChannel(this)

        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
        if (manager == null) {
            Log.e(TAG, "NotificationManager is null — cannot show notification")
            return
        }

        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(R.drawable.notification_sm_icon)
            .setContentTitle(title)
            .setContentText(body)
            .setStyle(
                NotificationCompat.BigTextStyle()
                    .bigText(body)
                    .setBigContentTitle(title)
            )
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setContentIntent(pendingIntent)

        val notificationId = (System.currentTimeMillis() % 100000).toInt()
        manager.notify(notificationId, builder.build())
        Log.d(TAG, "Notification displayed with ID: $notificationId on channel: $CHANNEL_ID")
    }

    private fun saveTokenLocally(token: String) {
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(TOKEN_KEY, token)
            .apply()
        Log.d(TAG, "FCM Token saved to SharedPreferences ✅")
    }

    private fun sendTokenToServer(token: String) {
        try {
            val postData = JSONObject().apply {
                put("userId", Constants.USER_ID)
                put("fcmToken", token)
                put("platform", "android")
            }

            WebServiceHelper.callHttpWebService(
                this,
                this,
                WebServiceHelper.RestMethodType.POST,
                Constants.Notification_Base_Url + "/register-token",
                "FCM_TOKEN_UPDATE",
                postData.toString()
            )

            Log.d("FCM", "FCM Token sent to backend ✅")
        } catch (e: Exception) {
            Log.e("FCM", "Failed to send FCM token to backend: ${e.message}")
            e.printStackTrace()
        }
    }

    override fun onClick(view: View) {}

    override fun onAsyncTaskComplete(httpResult: HttpResultDo) {
        try {
            if (httpResult.result == WebServiceHelper.ServiceCallStatus.Success) {
                val result = JSONObject(httpResult.responseContent ?: "")
                if ("FCM_TOKEN_UPDATE" == httpResult.requestType) {
                    val msg = result.optString("message")
                    Log.d("FCM_TOKEN_UPDATE", msg)
                }
            }
        } catch (e: Exception) {
            e.fillInStackTrace()
        }
    }

    companion object {
        private const val TAG = "FCM_Service"
        const val CHANNEL_ID = "lauditor_notifications_v3"
        private const val CHANNEL_NAME = "Lauditor Notifications"
        private const val PREFS_NAME = "MyPrefs"
        private const val TOKEN_KEY = "fcm_token"

        @JvmStatic
        fun initNotificationChannel(context: Context) {
            val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager ?: return
            try {
                manager.deleteNotificationChannel("lauditor_channel")
                manager.deleteNotificationChannel("lauditor_notifications_v2")
            } catch (e: Exception) {
                // ignore
            }

            if (manager.getNotificationChannel(CHANNEL_ID) == null) {
                val soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
                val audioAttributes = AudioAttributes.Builder()
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                    .build()

                val channel = NotificationChannel(
                    CHANNEL_ID,
                    CHANNEL_NAME,
                    NotificationManager.IMPORTANCE_DEFAULT
                ).apply {
                    description = "Push notifications for Lex-Z Lawyers App"
                    enableLights(false)
                    enableVibration(true)
                    setSound(soundUri, audioAttributes)
                    setShowBadge(true)
                    lockscreenVisibility = NotificationCompat.VISIBILITY_PUBLIC
                }
                manager.createNotificationChannel(channel)
                Log.d(TAG, "Notification channel created: $CHANNEL_ID ✅")
            }
        }


        @JvmStatic
        fun clearAllNotifications(context: Context) {
            val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
            manager?.let {
                it.cancelAll()
                Log.d(TAG, "All notifications cleared from notification center ✅")
            }
        }

        @JvmStatic
        fun logoutToken(context: Context, token: String) {
            try {
                val postData = JSONObject().apply {
                    put("fcmToken", token)
                }

                val dummyCallback = object : AsyncTaskCompleteListener {
                    override fun onClick(view: View) {}
                    override fun onAsyncTaskComplete(httpResult: HttpResultDo) {}
                }
                WebServiceHelper.callHttpWebService(
                    dummyCallback,
                    context,
                    WebServiceHelper.RestMethodType.POST,
                    Constants.Notification_Base_Url + "/logout-token",
                    "FCM_TOKEN_LOGOUT",
                    postData.toString()
                )

                Log.d("FCM", "Logout token sent to backend ✅")
            } catch (e: Exception) {
                Log.e("FCM", "Failed to send logout token to backend: ${e.message}")
            }
        }
    }
}
